### PR TITLE
Streaks system.

### PR DESCRIPTION
**Pull request by zyro**

### CODE REVIEW

File: hiro.proto [hunk 1]
```diff
@@ -2257,3 +2266,87 @@ message SyncResponse {
...
+  // Current overall progress count.
+  int64 count = 3;
+  // Maximum allowed overall progress count.
+  int64 max_count = 4;
+  // Progress count submitted during this reset.
+  int64 count_current_reset = 5;
+  // Maximum progress count allowed during any single reset.
+  int64 max_count_current_reset = 6;
+  // How much overall count is lost per idle reset.
+  int64 idle_count_decay_reset = 7;
```
Comment (adam-sutcliffe) [inline hiro.proto:2307]: Is the reset for when a user has passed the `reset_time_sec` without a login? So is `count_current_reset` the last streak they maintained before a reset? What is the `max_count_current_reset` for? and is the `idle_count_decay_reset` the value to remove on reset? If left empty does it do a full reset to zero?

File: hiro.proto [hunk 2]
```diff
@@ -2257,3 +2266,87 @@ message SyncResponse {
...
+// An individual streak, along with its status and progress if any.
+message Streak {
+  // Unique identifier.
+  string id = 1;
+  // Display name for this streak.
+  string name = 2;
+  // Current overall progress count.
+  int64 count = 3;
+  // Maximum allowed overall progress count.
+  int64 max_count = 4;
```
Comment (adam-sutcliffe) [inline hiro.proto:2301]: Can i put something like a `-1` for an infinite max count?

File: hiro.proto [hunk 3]
```diff
@@ -2257,3 +2266,87 @@ message SyncResponse {
...
+  int64 prev_reset_time_sec = 9;
+  // The UNIX time (for gRPC clients) or ISO string (for REST clients) when the streak's next reset period is.
+  int64 reset_time_sec = 10;
+  // The UNIX time (for gRPC clients) or ISO string (for REST clients) when the streak was first registered for this user.
+  int64 create_time_sec = 11;
+  // The UNIX time (for gRPC clients) or ISO string (for REST clients) when the streak was last updated.
+  int64 update_time_sec = 12;
+  // The UNIX time (for gRPC clients) or ISO string (for REST clients) when the streak was last claimed.
+  int64 claim_time_sec = 13;
+  // All rewards configured for this streak.
```
Comment (adam-sutcliffe) [inline hiro.proto:2320]: I'm also guessing these are all readonly properties too? As only the progress will be updated from the client via the update request call?

File: hiro.proto [hunk 4]
```diff
@@ -2257,3 +2266,87 @@ message SyncResponse {
...
+  // The UNIX time (for gRPC clients) or ISO string (for REST clients) when this reward was claimed.
+  int64 claim_time_sec = 4;
+}
+
+// An individual streak, along with its status and progress if any.
+message Streak {
+  // Unique identifier.
+  string id = 1;
+  // Display name for this streak.
+  string name = 2;
```
Comment (adam-sutcliffe) [inline hiro.proto:2297]: is it possible to also get a description as well as a name?

File: hiro.proto (hunk 1 above)
Comment (zyro) [reply hiro.proto:2307]: A reset decay happens when a full reset passes without the user having submitted any updates to this streak. The `count_current_reset` is the count submitted during the current reset only. `max_count_current_reset` is the maximum count that a user can submit during a single reset. `idle_count_decay_reset` is the value that is removed per missed reset (so missing 2 resets in a row would remove `2 * idle_count_decay_reset`) and if left empty indicates no decay.

File: hiro.proto (hunk 2 above)
Comment (zyro) [reply hiro.proto:2301]: Max count of `0` indicates no max count.

File: hiro.proto (hunk 3 above)
Comment (zyro) [reply hiro.proto:2320]: Correct, the update or reset calls are how the client submits data to streaks.

File: hiro.proto (hunk 4 above)
Comment (zyro) [reply hiro.proto:2297]: Will add. 👍

Comment (adam-sutcliffe) [reply hiro.proto:2297]: Thank you

File: hiro.proto (hunk 1 above)
Comment (adam-sutcliffe) [reply hiro.proto:2307]: cool, thanks for the clarification

**Pull request merged**